timApp/idesupport/utils.py

```diff
@@ -228,6 +232,85 @@ class TIDECourse:
     """
 
 
+def combine_code_blocks(program: str | None, by: str | None,
+                        by_code: str | None, filename: str | None) -> str:
+    """
+    Combine the code blocks.
+
+    :param program: Code block
+    :param by: Code block
+    :param by_code: Code block
+    :return: Combined code blocks
+    """
+    if program is None and by is None and by_code is None:
+        raise RouteException("No code found in the plugin")
+
+    if by is None and by_code is None:
+        return program
+
+    # Find the comment line syntax based on the file extension and create messages.
+    comment_line_syntax = find_comment_line_syntax(filename.split(".")[-1])
```

Comment:
Filename saattaa olla tässä vielä None. Tuolla jossain oli `ensure_filename`, mutta jos tätä idefile-luokkaa käytetään jossain muualla kuin tuossa yhdessä funktiossa niin ei oikein tiedä onko tuota aina kutsuttu.

Tätä vois refaktoroida niin että tuossa IdeFile-luokan alustuksessa tarkistettais tuo filename kuntoon ja samalla tehtäis tuo combine_code_blocks. Nää metodit  (ensure_filename ja combine_code_blocks) vois sitten siirtää suoraan tuonne IdeFile-luokkaan.

Jos haluaa tuon refaktoroinnin jättää myöhemmäks niin tässä varmaan voi heittää jonkun poikkeuksen jos filename on jäänyt Noneksi

Reply:
Menee hankalaksi laittaa esim. konstruktoriin tätä logiikkaa, mutta heitetään poikkeus filenamen puuttuessa.